core/src/test/java/org/opensearch/sql/utils/ComparisonUtil.java

```diff
@@ -11,45 +11,58 @@
+
   /**
    * Util to compare the object (integer, long, float, double, string) values.
-   * ExprValue A
+   * Allows comparing different datetime types and requires `FunctionProperties` object for that.
    */
-  public static int compare(ExprValue v1, ExprValue v2) {
+  public static int compare(FunctionProperties functionProperties, ExprValue v1, ExprValue v2) {
     if (v1.isMissing() || v2.isMissing()) {
       throw new ExpressionEvaluationException("invalid to call compare operation on missing value");
     } else if (v1.isNull() || v2.isNull()) {
       throw new ExpressionEvaluationException("invalid to call compare operation on null value");
+    } else if (v1.type() != v2.type() && v1.isDateTime() && v2.isDateTime()) {
+      return extractDateTime(v1, functionProperties)
+          .compareTo(extractDateTime(v2, functionProperties));
     }
+    return compare(v1, v2);
+  }
 
-    if (v1 instanceof ExprByteValue) {
-      return v1.byteValue().compareTo(v2.byteValue());
-    } else if (v1 instanceof ExprShortValue) {
-      return v1.shortValue().compareTo(v2.shortValue());
-    } else if (v1 instanceof ExprIntegerValue) {
-      return getIntegerValue(v1).compareTo(getIntegerValue(v2));
-    } else if (v1 instanceof ExprLongValue) {
-      return getLongValue(v1).compareTo(getLongValue(v2));
-    } else if (v1 instanceof ExprFloatValue) {
-      return getFloatValue(v1).compareTo(getFloatValue(v2));
-    } else if (v1 instanceof ExprDoubleValue) {
-      return getDoubleValue(v1).compareTo(getDoubleValue(v2));
-    } else if (v1 instanceof ExprStringValue) {
-      return getStringValue(v1).compareTo(getStringValue(v2));
-    } else {
+  /**
+   * Util to compare the object (integer, long, float, double, string) values.
+   */
+  public static int compare(ExprValue v1, ExprValue v2) {
+    if (v1.isMissing() || v2.isMissing()) {
+      throw new ExpressionEvaluationException("invalid to call compare operation on missing value");
+    } else if (v1.isNull() || v2.isNull()) {
+      throw new ExpressionEvaluationException("invalid to call compare operation on null value");
+    } else if (v1.type() != v2.type()) {
       throw new ExpressionEvaluationException(
+          "invalid to call compare operation on values of different types");
+    }
+
+    switch ((ExprCoreType)v1.type()) {
+      case BYTE: return v1.byteValue().compareTo(v2.byteValue());
+      case SHORT: return v1.shortValue().compareTo(v2.shortValue());
+      case INTEGER: return getIntegerValue(v1).compareTo(getIntegerValue(v2));
+      case LONG: return getLongValue(v1).compareTo(getLongValue(v2));
+      case FLOAT: return getFloatValue(v1).compareTo(getFloatValue(v2));
+      case DOUBLE: return getDoubleValue(v1).compareTo(getDoubleValue(v2));
+      case STRING: return getStringValue(v1).compareTo(getStringValue(v2));
+      case BOOLEAN: return v1.booleanValue().compareTo(v2.booleanValue());
+      case TIME: return v1.timeValue().compareTo(v2.timeValue());
```

Comment:
For time, shouldn't we potentially use datetimeValue() if the other type is DATETIME or timestampValue if the other type is TIMESTAMP? 
same for date...

Reply:
1. This function has `v1` and `v2` values of the same type, so we can compare time to time without overhead of producing and comparting timestamps.
2. We can't get timestamp/datetime from time without ~dark magic~ `FunctionProperties`